### DISCUSSION TITLE
Add support for PyBind11

### DISCUSSION THREAD

**Discussion by jmirabel** [stack-of-tasks/pinocchio#1518] [Ideas]
Hi,

for several projects, I made Python bindings with PyBind11 and I needed to be able to pass back and forth Pinocchio objects. I have a slightly different implementation in each project, which is not very clean.

As people usually prefers PyBind11 to Boost, I think it is reasonable to have in this project one header which contains everything one needs to do the conversions. Another reasonable option is to use a repo (e.g. pinocchio-pybind11) for this, although, if it is a single header, I'm not sure it's worth it.

I can implement this (with a low bandwidth but there is no emergency) if we agree (and if it does not exists already).

**Comment by jcarpent** [comment]
Indeed, it would be interesting. We already provide with EigenPy and Pinocchio-3x the conversion with SWIG objects.

**Comment by jcarpent** [comment]
If @jmirabel you can start providing your existing set of tools (with unit test), it will be easier for everyone to continue from that.

**Reply by jmirabel** [reply]
See PR #1519

**Comment by wxmerkt** [comment]
This would be very useful - I've also created converters for some Pinocchio objects (Model, Data, Frame, Force, etc. - but not all) and successfully passed them to Pybind11-wrapped modules and retrieved Pinocchio objects back from them. However, there are no unittests. If of use, happy to provide them, but they may also be the same you as yours. I used a macro to implemente the `type_caster` which then only requires a one-line instantiation per Pinocchio object. It's about 42 lines overall

**Reply by jcarpent** [reply]
@jmirabel @wxmerkt Both contributions will be highly appreciated. 
I let you decide which one of you open the PRs on top of which we may build a consensus.

@wxmerkt Does your solution is copy-free?

**Reply by jmirabel** [reply]
Mine is definitely longer but it allows to bind the arguments and return value almost automatically in some cases.
